### PR TITLE
Fix Application instances created multiple times

### PR DESCRIPTION
**Pull request by ChristophWurst**
Apps can have an `Application`, which is loaded automatically for them. This is used for initialization.

However, we didn't pay attention to the number of instances we create. Some requests had one, some had two. This ended in stuff being run twice, like event listeners firing off twice due to double registration. A workaround was added to Mail, Suspicious Login and Push, but it's time to fix this here.

So, here it is. I changed the code to always fetch the Application instance from the server container (except for when we're inside the container code, then we have no other choice but create the instance manually). I tested the code with Mail, where I attached a debugger to the BootstrapSingleton and the instance was never fetched a second time.

We can't solve the problem for apps that still create an instance of the class manually. But for those where it's a problem, they can now safely use the server container to get an instance, e.g. in app.php or routes.php.

You can review the commits separately. The first one changes the way server creates instances, the second one fixes all other usages.

In theory we could even have a tiny performance improvement as fewer class instances are loaded. But I double that it makes a noticeable difference.

### CODE REVIEW

**Review by skjnldsv — APPROVED** [review]
Shall we document it so that all apps migrate to this new syntax?

**Comment by ChristophWurst** [comment]
> Shall we document it so that all apps migrate to this new syntax?

As a recommendation, yeah. Because when you run into the troubles they aren't so obvious.

File: apps/encryption/appinfo/routes.php [hunk 1]
```diff
@@ -24,7 +24,9 @@
 
 namespace OCA\Encryption\AppInfo;
 
-(new Application())->registerRoutes($this, array('routes' => array(
+/** @var Application $app */
+$app = \OC::$server->query(Application::class);
+$app->registerRoutes($this, array('routes' => array(
```
Comment (nickvergessen) [inline apps/encryption/appinfo/routes.php]: Just return the routes as an array in this file

File: apps/files/appinfo/routes.php [hunk 2]
```diff
@@ -29,7 +29,8 @@
  */
 namespace OCA\Files\AppInfo;
 
-$application = new Application();
+/** @var Application $application */
+$application = \OC::$server->query(Application::class);
 $application->registerRoutes(
```
Comment (nickvergessen) [inline apps/files/appinfo/routes.php]: Same here

File: apps/files_trashbin/appinfo/routes.php [hunk 3]
```diff
@@ -24,7 +24,8 @@
 
 namespace OCA\Files_Trashbin\AppInfo;
 
-$application = new Application();
+/** @var Application $application */
+$application = \OC::$server->query(Application::class);
 $application->registerRoutes($this, [
```
Comment (nickvergessen) [inline apps/files_trashbin/appinfo/routes.php]: And here

File: apps/files_versions/appinfo/routes.php [hunk 4]
```diff
@@ -26,7 +26,8 @@
 
 namespace OCA\Files_Versions\AppInfo;
```
Comment (nickvergessen) [inline apps/files_versions/appinfo/routes.php]: Also these namespaces are invalid. That namespace lives in `lib/AppInfo`

File: apps/files/appinfo/routes.php (hunk 2 above)
Comment (rullzer) [reply apps/files/appinfo/routes.php]: here it is beause of the shitty routes below which we should get rid of at some point

**Review by rullzer — APPROVED** [review]

**Comment by skjnldsv** [comment]
Please rebase autosquash :)

**Comment by rullzer** [comment]
Somehow CI is not happy

**Comment by ChristophWurst** [comment]
🤯

Bildschirmfoto von 2019-10-21 09-43-02

**Comment by rullzer** [comment]
Yeah best thing is I can't trigger it locally...

**Comment by icewind1991** [comment]
Fixed the trashbin tests

**Pull request merged**